### PR TITLE
[BREAKING] Fix sanitizer tests.

### PR DESCRIPTION
**Pull request by trivialfis**
There's another one in thread iter I don't know how to fix yet.

### CODE REVIEW

**Comment by trivialfis** [comment]
@hcho3 Do you have any update on the thead sanitizing front?

**Comment by trivialfis** [comment]
Okay, I made some changes to thread utilities, can't reproduce any sanitizer errors on my local machine now (by running unittests a dozen times).

**Comment by trivialfis** [comment]
Oh no... There are others.

**Comment by trivialfis** [comment]
This doesn't resolve https://github.com/dmlc/xgboost/issues/4619 , only eliminating tsan errors in dmlc_unit_tests.

**Comment by trivialfis** [comment]
Done.  Ping @hcho3

File: include/dmlc/threadediter.h [hunk 1]
```diff
@@ -138,11 +138,9 @@ class ThreadedIter : public DataIter<DType> {
    * \brief initialize the producer and start the thread
    *   can only be called once
    * \param producer pointer to the producer
-   * \param pass_ownership whether pass the ownership to the iter
-   *    if this is true, the threaditer will delete the producer
-   *    when destructed
    */
-  inline void Init(Producer *producer, bool pass_ownership = false);
+  inline void Init(std::shared_ptr<Producer> producer);
+  // inline void Init(std::unique_ptr<Producer> producer);
```
Comment (hcho3) [inline include/dmlc/threadediter.h]: Do we need this comment?

**Comment by trivialfis** [comment]
Will remove that comment, missed.  I made the breaking change out of good heart ... Even the test can implicitly capture a producer in lambda that goes out of scope sooner than threadefiter, I think others can get it wrong very easily.

**Comment by hcho3** [comment]
@trivialfis Got it. I marked this PR as breaking, since `ThreadedIter::Init()` now takes in a `shared_ptr`.

**Pull request merged**